Master2

bilalSerwan opened the pull request.

bilalSerwan committed:
create zekir app

bilalSerwan committed:
some changes

bilalSerwan committed:
Refactor The Code

bilalSerwan committed:
readme file

RoonyShekho09 commented:
The title on the top bar appears to be cut, make sure the text is showing on different screen sizes

<img src="https://github.com/user-attachments/assets/6807fa38-c4a5-4a6e-b648-09205ffdeae2" width="400" height="790">

RoonyShekho09 commented:
Having a Toast covering a button is not a good UX, also the BottomBar is not at the bottom.

<img src="https://github.com/user-attachments/assets/d075b61e-79ce-4d71-b062-a9984ca8422b" width="400" height="790">

RoonyShekho09 commented:
The Toast is not showing the numbers correctly

[Screen_recording_20240812_102324.webm](https://github.com/user-attachments/assets/330a3bea-4c35-4bd5-b38f-423df5bedc5b)

RoonyShekho09 commented:
I will review when the points are fixed.

bilalSerwan commented:
> I will review when the points are fixed.

thank you for reviewing inshallah i will fix it

bilalSerwan committed:
fix kak Roony points

RoonyShekho09 commented:
1. For future commits, make sure the commit message describe what you did. If you fixed something, write what you fixed in what context, for example, "fix toast message when the user clicks on the play button" or "implement permission," so when someone reads the commit message, they will know what has changed. 
2. The branch name should also be related to what you are doing. Here you have master2, which doesn't mean anything. It's enough to have one branch named "main". Both "master" and "main" mean the same thing, which is your base branch. Others will be your feature branches, so if you want to implement screen called "A" the branch would be "implement A screen".

bilalSerwan commented:
> 1. For future commits, make sure the commit message describe what you did. If you fixed something, write what you fixed in what context, for example, "fix toast message when the user clicks on the play button" or "implement permission," so when someone reads the commit message, they will know what has changed.
> 2. The branch name should also be related to what you are doing. Here you have master2, which doesn't mean anything. It's enough to have one branch named "main". Both "master" and "main" mean the same thing, which is your base branch. Others will be your feature branches, so if you want to implement screen called "A" the branch would be "implement A screen".

inshallah for Future commits and Projects I will do it, and Thank you for everything

bilalSerwan committed:
Reformating the code by uinge shortcut Alt+Ctrl+L for All Files, and adding (No Favorite Categories Yet) Text to Favorite Screen if the user don't have any favorite zekirs

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/MainActivity.kt:
```diff
@@ -0,0 +1,89 @@
+
+
+val LocalNavController = compositionLocalOf<NavHostController> { error("error in navController") }
+val LocalViewModel = compositionLocalOf<MainViewModel> { error("error in view model") }
```
Use [Hilt](https://developer.android.com/training/dependency-injection/hilt-android) instead of `compositionLocalOf` to inject the `ViewModel`.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/MainActivity.kt:
```diff
@@ -0,0 +1,89 @@
+
+val LocalNavController = compositionLocalOf<NavHostController> { error("error in navController") }
+val LocalViewModel = compositionLocalOf<MainViewModel> { error("error in view model") }
+val LocalModifier = compositionLocalOf<Modifier> { error("error in modifier") }
```
There is no point to provide the `Modifier` using `compositionLocalOf` this is wrong, the `Modifier` should be passed as an argument to other compostable functions.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/MainActivity.kt:
```diff
@@ -0,0 +1,89 @@
+        enableEdgeToEdge()
+        setContent {
+            ZekrAppTheme {
+                val toast: Toast = Toast.makeText(this, "", Toast.LENGTH_SHORT)
```
I don't like this

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/MainActivity.kt:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+    private fun showToast(toast: Toast, message: String) {
```
use `SnackBar` instead of `Toast`.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/MainActivity.kt:
```diff
@@ -0,0 +1,89 @@
+            Navigation()
+        }
+    )
+}
```
This should be refactored.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/data/data/CategoryData.kt:
This should be as a `Json` or `CSV` then it parsed and mapped as a model.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/data/data/CategoryData.kt:
why the package name is `data` inside another package called `data`?
`data/data` this is bad organizating.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/data/data/zekrdata.kt:
This should be as a `Json` or `CSV` then it parsed and mapped as a model.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/data/model/CategoryModel.kt:
```diff
@@ -0,0 +1,16 @@
+package com.fastlink.zekrapp.data.model
+
+//@Entity(tableName = "zekr_category_table")
```
Don't keep the dead code.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/data/model/CategoryModel.kt:
```diff
@@ -0,0 +1,16 @@
+    val name: String,
+//    val zekrs: List<Zekr>,
+//    @ColumnInfo(name = "name_abstract")
+    val nameAbstract:String,
```
This is not descriptive naming.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/data/model/CategoryModel.kt:
```diff
@@ -0,0 +1,16 @@
+//    @ColumnInfo(name = "name_abstract")
+    val nameAbstract:String,
+//    @ColumnInfo(name="bookmarked")
+    var isFavorite : Boolean =false
```
You should be aware of code formatting.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/data/model/CategoryModel.kt:
```diff
@@ -0,0 +1,16 @@
+    var isFavorite : Boolean =false
+)
+
+
```
one empty line in the end of file is enough.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/data/model/ZekirModel.kt:
```diff
@@ -0,0 +1,19 @@
+package com.fastlink.zekrapp.data.model
+
+//@Entity(tableName = "zekr_table")
```
More of dead code.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/presentation/HomeScreen.kt:
```diff
@@ -0,0 +1,41 @@
+            .padding(
+                top = 90.dp,
+                bottom = 100.dp,
+            )
```
This is very large padding, you can use `Box` to center the content then use `offset` modifier to adjust the place.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/presentation/utils/bottomAppBar/zekirScreenBottomBar.kt:
```diff
@@ -0,0 +1,106 @@
+
+    BottomAppBar(
+        modifier = Modifier,
+//            .padding(bottom = if (context.hasNavigationBar()) 40.dp else 0.dp),
```
Dead code.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/presentation/utils/bottomAppBar/homeAndFavoriteBottomBar.kt:
```diff
@@ -0,0 +1,57 @@
+        contentPadding = PaddingValues(horizontal = 90.dp)
+    ) {
+        BottomAppBarItem(icon = Icons.Default.Home,
+            text = "Home",
```
Hardcoded, the string resource should be extracted.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/presentation/utils/bottomAppBar/homeAndFavoriteBottomBar.kt:
```diff
@@ -0,0 +1,57 @@
+        Spacer(modifier = Modifier.weight(1f))
+        BottomAppBarItem(
+            icon = Icons.Default.Favorite,
+            text = "Favorite",
```
Hardcoded, the string resource should be extracted.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,198 @@
+                style = MaterialTheme.typography.titleMedium,
+            )
+            Text(
+                text = "(${if (zekirs[zekirNumber.intValue].counter == "null") "مرة واحدة" else zekirs[zekirNumber.intValue].counter})",
```
Hardcoded, the string resource should be extracted.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/data/data/Categories.kt:
this can be moved to a singleton

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/data/data/Zekrs.kt:
this can be moved to a singleton

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/presentation/FavoriteScreen.kt:
```diff
@@ -0,0 +1,60 @@
+                        .background(color = MaterialTheme.colorScheme.background)
+                )
+                Text(
+                    text = "No Favorite Categories Yet",
```
Hardcoded, the string resource should be extracted.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/presentation/FavoriteScreen.kt:
```diff
@@ -0,0 +1,60 @@
+        modifier = modifier
+            .padding(
+                top = 90.dp,
+                bottom = 100.dp,
```
This is very large padding; you can use Box to center the content then use offset modifier to adjust the place.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/presentation/FavoriteScreen.kt:
```diff
@@ -0,0 +1,60 @@
+                        .size(200.dp)
+                        .background(color = MaterialTheme.colorScheme.background)
+                )
+                Text(
```
This is wrong, you can use `Box` to center the content.

Al-Taie commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,198 @@
+                    rememberScrollState()
+                )
+                .clickable(
+                    enabled = (zekirCounter.intValue < zekirs[zekirNumber.intValue].counterNumber) || zekirNumber.intValue < zekirs.size - 1,
```
this is complex

Al-Taie reviewed (changes requested):
Good job, I leaved some comments.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,198 @@
+    val scope = rememberCoroutineScope()
+    val category = viewModel.getCategoryById(categoryId)
+    val zekirs = viewModel.getZekirByCategoryId(category.id)
+    val zekirNumber = remember {
```
Why this also is not in your view model?

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,198 @@
+    val zekirNumber = remember {
+        mutableIntStateOf(0)
+    }
+    val zekirCounter = remember {
```
This also could be moved to the view model

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,198 @@
+        animationSpec = tween(1000)
+    )
+
+    fun onClicked() {
```
Here if you read this function name you don't know what it's doing, clicked on what? Also you should make the composable for showing the ui, handling the logic and state management should be in the view model

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,198 @@
+                .verticalScroll(
+                    rememberScrollState()
+                )
+                .clickable(
```
Same as Ahmed mentioned. You should try to make the ui part only for showing, and leave the logic for the view model.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/data/data/CategoryData.kt:
```diff
@@ -0,0 +1,935 @@
+package com.fastlink.zekrapp.data.data
```
You can call it repository instead of data

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/utils/bottomAppBar/zekirScreenBottomBar.kt:
```diff
@@ -0,0 +1,106 @@
+import com.fastlink.zekrapp.data.model.CategoryModel
+
+
+//this function is used to check if the device has a navigation bar or not
```
That's already clear from the function name, no need for redundant comments.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/utils/bottomAppBar/zekirScreenBottomBar.kt:
```diff
@@ -0,0 +1,106 @@
+
+
+//this function is used to check if the device has a navigation bar or not
+//hint: i didn't write this function, i found it on the internet
```
You don't have to mention that you didn't write it.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/data/data/CategoryData.kt:
```diff
@@ -0,0 +1,935 @@
+package com.fastlink.zekrapp.data.data
+
+val categoryData : ArrayList<ArrayList<Any>> = arrayListOf(
```
Naming it zekirType or zekirCategory is clearer than having just category, if we read it as just category we don't know what it is.

RoonyShekho09 replied:
Instead of having array list of array list you could have a data class representing an item.

RoonyShekho09 commented:
Also the application is dark in one screen, and light in another. This doesn't make it look good.

bilalSerwan committed:
Refactoring the code and extract Texts from ui and add to string file in res/values and refactoring MainActivity file and delete dead Codes or comments and changing varibale name and chaning the bottomAppBar of ZekirScreen and changing name of data folder to appData

bilalSerwan committed:
Add a halfCircle function to the Modifier and refactor the Zekir screen by extracting logic from the UI

bilalSerwan commented:
Regarding the last commit, I didn't complete extracting logic from Ui in Zekir Screen because we will switch our laptops with Xatw Sawin for A week, And Maybe I Can't work on this PR or Project for 2 or 3 days

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/appData/data/Zekrs.kt:
```diff
@@ -9,10 +10,10 @@ fun getZekrs(){
+                plainZekirTitle = i[3].toString(),
                 hint = i[4].toString(),
-                counter = i[5].toString(),
+                counterAsString = i[5].toString(),
```
Counter is enough, no need to add asString to it

bilalSerwan replied:
We have two counters for  zikr recitation: one counter records the number of recitations numerically, while the other counter represents the number  as a textual string
for example 
this is counterASString Varibal
<img width="359" alt="Screenshot 2024-08-15 at 2 17 35 PM" src="https://github.com/user-attachments/assets/dc0b47a9-10d5-4394-a900-ca62e232dc4a">

and this is counter 
<img width="417" alt="Screenshot 2024-08-15 at 2 18 18 PM" src="https://github.com/user-attachments/assets/6c4b2392-0597-4561-9e87-3360fdb247c7">

RoonyShekho09 replied:
But numbers can be stored in a string too, name them _numericCounter_ and _textCounter_.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/appData/model/CategoryModel.kt:
```diff
@@ -0,0 +1,8 @@
+data class CategoryModel(
+    val id: Int,
+    val categoryTitle: String,
+    val plainCategoryTitle: String,
```
What does plainCategoryTitle mean, and why is it here if it's not used?

bilalSerwan replied:
we use it but not now, plainCategoryTitle represents the category title without diacritical marks (i.e., without kasra, fatha, or other vowel signs) , if we add search functionality we use this plain Category title

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/viewModel/ViewModel.kt:
```diff
@@ -0,0 +1,87 @@
+                    if (zekirNumber.intValue < zekirs.size - 1) {
+                        zekirCounter.intValue = 0
+                        zekirNumber.intValue++
+//                        scope.launch {
```
Ahmed already mentioned dead code.

bilalSerwan replied:
sorry for dead code i will remove it, kak ahmad mentioned to use `SnackBar` instead of `Toast`

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -35,77 +34,58 @@ import androidx.compose.ui.res.stringResource
                 FloatingActionButton(
                     onClick = {
-                        onClicked()
+                        viewModel.onClicked(scope)
```
As I mentioned before, onClicked is not enough, because here you have a lot of components, you should mention clicked on what.

bilalSerwan replied:
nothing comes to my mind for naming this function because we use it in 2 places when click on `FAB` and when clicked on the zekir part 
both place clickable and both do the same action
<img width="422" alt="Screenshot 2024-08-15 at 2 39 15 PM" src="https://github.com/user-attachments/assets/cfcec7ab-f8d6-41a9-be51-e1720d4f0f9d">

RoonyShekho09 replied:
Make two functions for onClick, and name them according to where they are clicked. In the future maybe you will have different logic depending on whether you clicked on the button or the card, if you have only one function then the logic will be complex, but if you have two of them the logic will be separate for each one.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/appData/data/Categories.kt:
```diff
@@ -9,8 +10,8 @@ fun getCateories(){
         categoriesData.add(
             CategoryModel(
```
I already mentioned the name of the class, category is not clear, category of what?

bilalSerwan replied:
Sorry About that i will change it soon

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/utils/bottomAppBar/homeAndFavoriteBottomBar.kt:
```diff
@@ -12,43 +12,36 @@ import androidx.compose.runtime.Composable
 
 @Composable
-fun BottomBar() {
+fun BottomAppBarForHomeAndFavoriteScreens(
```
Choose a better name, don't mention where the component is used, suppose you changed your mind and wanted to use the component in a different screen, are you going to change the name and add another "and"?

bilalSerwan replied:
you are right, but we have 2 different BottomAppBars we can't say BottomAppBar1 And BottomAppBar2 for that, 
we have a only 3 screens if we had another screen I didn't change the name this  i used the same previous name `BottomBar`

RoonyShekho09 replied:
Either delete one of them and make the bottom bar reusable for all the screens, or choose better names for them. If we have 2 bottom bars that doesn't mean to name them like that.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/appData/data/zekrdata.kt:
```diff
@@ -1,4 +1,4 @@
-package com.fastlink.zekrapp.data.data
+package com.fastlink.zekrapp.appData.data
```
"appData" is not better than "data", make the package data then inside it repository. Or just put everything inside one package called "data". 

Take a look at this:
https://github.com/android/compose-samples/tree/main/JetLagged/app/src/main/java/com/example/jetlagged/data

bilalSerwan replied:
Thank you for providing this resource; I appreciate it. Please note that I will delete this folder at the end of the project. Initially, we will retrieve the data from a local database, but I included this folder to facilitate the UI implementation. Kak Ahmad instructed me to focus solely on creating the UI and not to handle the data until the roadmap is completed. While this is not ideal, I will make an effort to manage it. For the time being, I am considering using a CSV file as an alternative to the data currently stored in the ArrayList

RoonyShekho09 replied:
I'm only talking about the package names.

RoonyShekho09 reviewed (changes requested):
I wrote some new comments, but a lot of the things are still the same and haven't changed from the old comments, make sure to read all the comments.

bilalSerwan commented:
> I wrote some new comments, but a lot of the things are still the same and haven't changed from the old comments, make sure to read all the comments.

inshallah i will do it

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/utils/bottomAppBar/zekirScreenBottomBar.kt:
```diff
@@ -0,0 +1,81 @@
+) {
+    val viewmodel = LocalViewModel.current
+
+    BottomAppBar(
```
You have BottomNavigation, you don't need BottomAppBar.

RoonyShekho09 commented:
Remove these white spaces

<img src="https://github.com/user-attachments/assets/25a7a711-8a43-4643-bc7d-e62c5f00620a" width="400" height="790">

Commit:
create singleton object to category and zekir, Refactoring AppBar And Bottom Bar for All Screens , adding  CSV file of Zekir Category data but I tried to use CSv file to read Zekir Data but have some issue in the csv file when exporting in sqlite database file , after Refactoring deleted some unneccary files

Commit:
adding snackbar and create Repository file to get data form file or database if we have a local database in future

bilalSerwan commented:
I didn't know why the code regarding the last commit wasn't committed. I did it quickly because Xatw Sawin needed her laptop for something, so I committed and pushed, but I didn't understand why the code in the Repository file and the Snackbar didn't get included. Sorry about that inshallah after xatw sawin completed her work i will push  it again

bilalSerwan commented:
If I'm not mistaken, only Kak Ahmad's point remains, which is to use Hilt instead of Local Composition to inject the ViewModel. I searched for it yesterday but didn't understand it; inshallah, I will research it further

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/viewModel/ViewModel.kt:
```diff
@@ -0,0 +1,79 @@
+    private var _zekirCategories = mutableStateOf(ZekirCategorySingleton.zekirCategoriesData)
+    val zekirNumber = mutableIntStateOf(0)
+    val zekirCounter = mutableIntStateOf(0)
+    var zekirs :  List<ZekirModel> = emptyList()
```
Don't expose mutable variables to other classes, you don't want to make other classes to be able to modify your variables.

bilalSerwan replied:
i changed all this things but I didn't know why didn't come

bilalSerwan replied:
deleted both variables _zekirCategory and zekirCategory  instead of this i created a method getAllCategories and get data from singleton

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/viewModel/ViewModel.kt:
```diff
@@ -0,0 +1,79 @@
+            }
+        }
+    }
+    fun onClickedZekirCard(scope : CoroutineScope) {
```
If you have the same logic for onClickedFAB and onClickedZekirCard then extract the logic to a separate function and use it in both places.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,211 @@
+    val navController = LocalNavController.current
+    val clipboardManager = LocalClipboardManager.current
+    val scope = rememberCoroutineScope()
+    remember { viewModel.getZekirByCategoryId(categoryId) }
```
Why is this wrapped in remember? There are reasons to use remember you can read this:
https://developer.android.com/develop/ui/compose/state

Also no need to call getZekirByCategoryId here, you can simply call it in the view model when it's initialized.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,211 @@
+    val clipboardManager = LocalClipboardManager.current
+    val scope = rememberCoroutineScope()
+    remember { viewModel.getZekirByCategoryId(categoryId) }
+    val category = viewModel.getCategoryById(categoryId)
```
For any function that needs to be called when the screen is started you can call them in the view model init block.

RoonyShekho09 replied:
Also what is the difference between getZekirByCategoryId and getCategoryById? The second one should have a clearer name.

RoonyShekho09 replied:
What is getCategoryById? The name should be clearer.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,211 @@
+        animationSpec = tween(1000)
+    )
+
+    fun buttonIsEnable(): Boolean {
```
No point for this function to be here, you can place it in your view model.

RoonyShekho09 replied:
@bilalSerwan This hasn't changed yet. Please request review only when you're finished.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,211 @@
+            )
+            Text(
+                text = "\n(${
+                    if (viewModel.zekirs[zekirNumber.intValue].textCounter == "null") stringResource(
```
Why would you have "null" as a string?

bilalSerwan replied:
I explained that we have two fields or instances: `textCounter `and `numericCounter`. In the Zekir database, we have a column with two types of data: if Zekir is read once, the data is `null`, but if it is read more than once, the data is a `String`. I store this data in an ArrayList of type `Any`, which allows you to save either `null `or a `String`. However, this gives a warning because Any doesn't match the type `String?.` To resolve this, I store the data as a String by calling the `.toString` function when saving the data in the model. This way, I can check if the data is a null string, and if it is, I use a String Resource; if it's not null, I use `textCounter`

RoonyShekho09 replied:
I don't really want the explanation, when I ask why I'm only pointing out the point. Having it empty is better, null makes sense only as a value, not in a string.

bilalSerwan replied:
you are right, if I checked in first maybe is better 
![image](https://github.com/user-attachments/assets/855b9835-9933-4e7d-b9cd-d1bc16c603d9)
in the singleton file 
instead of the previous code I change it to this 
`if(i[5].toString()=="null") null else i[5].toString()`
and in the ZekirScreen Refactor the code to 
`viewModel.zekirs[zekirNumber].textCounter ?: stringResource(the text)`

What do you think, is this better or not?

bilalSerwan replied:
> you are right, if I checked in first maybe is better ![image](https://private-user-images.githubusercontent.com/121394643/359825134-855b9835-9933-4e7d-b9cd-d1bc16c603d9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQyMzA4NzksIm5iZiI6MTcyNDIzMDU3OSwicGF0aCI6Ii8xMjEzOTQ2NDMvMzU5ODI1MTM0LTg1NWI5ODM1LTk5MzMtNGU3ZC1iOWNkLWQxYmMxNmM2MDNkOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyMVQwODU2MTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kY2EwN2YwZTQ2ZmViZDJmOGIyZTVkMzA2MjhjOTdjYTlmMjQzMjFhNzE3NzM3YWNlYjVjMWVjODA4YTEwZDAzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.fK7z_IyrP2XaA5O9TMtFuECF_-9oOsC41llhx3Miayc) in the singleton file instead of the previous code I change it to this `if(i[5].toString()=="null") null else i[5].toString()` and in the ZekirScreen Refactor the code to `viewModel.zekirs[zekirNumber].textCounter ?: stringResource(the text)`
> 
> What do you think, is this better or not?

RoonyShekho09 replied:
You are still checking null inside a string, so nothing's changed.

bilalSerwan replied:
the data like this 
![image](https://github.com/user-attachments/assets/0c8f387c-3c32-4576-92f8-c577ab086c4b)
![image](https://github.com/user-attachments/assets/7caa9b1a-e6e8-4118-bd09-d509f7d8d106)
What is your suggestion about this? instead of checking how to handle this?

RoonyShekho09 replied:
First of all I don't know why the type is Any instead of having a data class representing the data. You can have your data like this:

`val zekirData: ArrayList<ZekirModel> = arrayListOf(`

bilalSerwan replied:
you are right I didn't thought that I change it to that inshallah thank you

Commit:
adding snackbar and create Repository file to get data form file or database if we have a local database in future

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/appData/category-repository.kt:
Don't name files this way, name this and the others same as the class name inside the file, "CategoryRepository" and not "category-repository."

RoonyShekho09 replied:
Plus you have repository and singleton. You don't need two separate classes for doing the same thing.

RoonyShekho09 replied:
Also don't leave anything named "category" specify what category you mean, everywhere not only in some places.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/appData/zekir_singleton.kt:
```diff
@@ -0,0 +1,20 @@
+        getZekrs()
+    }
+
+    private fun getZekrs() {
```
Are you getting or adding?

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,201 @@
+
+
+    LaunchedEffect(key1 = zekirNumber.intValue) {
+        snackBarState.showSnackbar("")
```
Why would you need to show an empty message? Avoid this.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,201 @@
+                .fillMaxSize()
+                .verticalScroll(rememberScrollState())
+                .clickable(
+                    enabled = buttonIsEnable(),
```
isButtonEnabled. Also don't leave it as button, specify which button.

Commit:
First of all, we made several changes: refactoring some file names, deleting the category-repository.kt file, updating the CategorySingleton file, separating ZekirScreen into composables, and splitting the view models into ZekirCategoryViewModel and ZekirViewModel

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/MainActivity.kt:
```diff
@@ -0,0 +1,50 @@
+        enableEdgeToEdge()
+        setContent {
+            val context = LocalContext.current
+            ZekirCategorySingleton.getZekirCategoriesFromCSVFile(context)
```
Call this in the view model upon initialization

bilalSerwan replied:
but i have one question you can't call this function in viewModel because need @Composable annotation for getting LocalContext

bilalSerwan replied:
you can remove this parametter and set the getZekirCategoryFromCSVFile fun to @Composable function but then you can't access it in init of viewmodel

bilalSerwan replied:
you must do that 
1. first you add annotation `Composable` to fun and removing parameter context and set variable in function that can access it 
2. create a fun in viewmodel and call `getZekirCategoryFromCSVFIle` on it 
3.call viewmodel function in initialization of the app 


that's true , do this ?

RoonyShekho09 replied:
LocalContext is not the only way to get the context.

bilalSerwan replied:
i solved this problem by another way

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/viewModel/ZekirViewModel.kt:
```diff
@@ -0,0 +1,50 @@
+
+class ZekirViewModel(categoryId: Int) : ViewModel() {
+    val zekirCounter = mutableIntStateOf(0)
+    var zekirs: List<ZekirModel> = emptyList()
```
Don't expose mutable variables publicly. You can do this to prevent other classes from modifying your variable:

```
var zekirs: List<ZekirModel> = emptyList()
                private set
```

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/viewModel/ZekirViewModel.kt:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    private fun getZekirByCategoryId(categoryId: Int) {
+        resetZekirCounter()
```
the function name is called getZekir... but here you are doing something else which is resetting. Move this to the init block. Because your function name should be what the function is doing.

bilalSerwan replied:
you are right i will change it now

bilalSerwan replied:
i didn't notate this

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/viewModel/ZekirViewModel.kt:
```diff
@@ -0,0 +1,50 @@
+        zekirCounter.intValue = 0
+    }
+
+    fun cardButtonIsEnable(zekirNumber: Int): Boolean {
```
isCardButtonEnabled

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,121 @@
+    val scaffoldState = rememberScaffoldState()
+    val navController = LocalNavController.current
+    val clipboardManager = LocalClipboardManager.current
+    val zekirCategory = categoryViewModel.getCategoryById(categoryId)
```
call this in the view model

bilalSerwan replied:
i didn't understand you ? what i do ?

RoonyShekho09 replied:
Call categoryViewModel.getCategoryById(categoryId) in the view model.

bilalSerwan replied:
i changed the name of function  to `getZeikrCategoryById` and should I call function in  View Model ?

bilalSerwan replied:
Kak @RoonyShekho09 
I made a mistake in the code by creating an instance in the ZekirCategoryViewModel class.
```kotlin
 val categoryViewModel: ZekirCategoryViewModel = viewModel()
```  
but we don't need a new instance then i changed to this 
```kotlin
 val categoryViewModel = LocalViewModel.current
```

RoonyShekho09 replied:
As I said before, LocalComposition is not for everything. For now use viewModel() until you read about Hilt.

bilalSerwan replied:
i read abou it but i didn't understand it  inshallah today i will search for it

RoonyShekho09 replied:
This is what the documentation says:

> Avoid CompositionLocal for concepts that aren't thought as tree-scoped or sub-hierarchy scoped. A CompositionLocal makes sense when it can be potentially used by any descendant, not by a few of them.

Also:

> An example of a bad practice is creating a CompositionLocal that holds the ViewModel of a particular screen so that all composables in that screen can get a reference to the ViewModel to perform some logic. This is a bad practice because not all composables below a particular UI tree need to know about a ViewModel.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,121 @@
+        pageCount = { zekirViewModel.zekirs.size },
+    )
+    val scaffoldState = rememberScaffoldState()
+    val navController = LocalNavController.current
```
LocalComposition is not used for everything. Read the documentation to see when you should use it:

https://developer.android.com/develop/ui/compose/compositionlocal

Here you can use rememberNavController instead.

bilalSerwan replied:
but if you change to this , created some issues

bilalSerwan replied:
the bottom App Bar don't show

RoonyShekho09 replied:
Then see why is this happening.

RoonyShekho09 replied:
Change this everywhere not only here.

bilalSerwan replied:
in this code 
`val navController = rememberNavController()
    val navBackStackEntry by navController.currentBackStackEntryAsState()`
returning the backstackEntry as null

RoonyShekho09 replied:
But don't call rememberNavController every time, pass the navController to the parameter of every screen and then initialize with rememberNavController in your activity.

bilalSerwan replied:
and navigation between pages didn't work ?

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/viewModel/ZekirViewModel.kt:
```diff
@@ -0,0 +1,50 @@
+        return zekirNumber + 1 < zekirs.size || zekirCounter.intValue < zekirs[zekirNumber].numericCounter
+    }
+
+    suspend fun handleFABClick(scope: CoroutineScope, zekirNumber: Int) =
```
ViewModel has its own scope, no need to pass the scope.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/viewModel/ZekirViewModel.kt:
```diff
@@ -0,0 +1,50 @@
+    suspend fun handleFABClick(scope: CoroutineScope, zekirNumber: Int) =
+        incrementZekirCounter(scope, zekirNumber)
+
+    suspend fun handleZekirCardClick(scope: CoroutineScope, zekirNumber: Int) =
```
Same here, don't pass the scope.

bilalSerwan replied:
i did it thank you i use ViewModelScope instead this coroutinescope

Commit:
I deleted the CoroutineScope parameter and used viewModelScope instead, renamed functions, modified ZekirCategoryViewModel, added a ZekirCategoryViewModelFactory file, and passed NavController as a parameter to all composables that use it, instead of using compositionLocalOf

Commit:
using hilt(DI) and fixing kak ronny's Points

bilalSerwan commented:
Until now, I’ve understood 50% about Hilt, how it works and how to implement it. However, there are still some things I need to learn, like using Hilt with fragments, navigation, and other components. So far, I’ve implemented it in this app, and the app works well without any bugs or issues. But I’m not sure if I’ve implemented it correctly.

RoonyShekho09 commented:
This [commit](https://github.com/bilalSerwan/ZekirApp/pull/3/commits/66424232b4da2c094803320883840e20ee19ddef) is empty.

RoonyShekho09 commented:
> Until now, I’ve understood 50% about Hilt, how it works and how to implement it. However, there are still some things I need to learn, like using Hilt with fragments, navigation, and other components. So far, I’ve implemented it in this app, and the app works well without any bugs or issues. But I’m not sure if I’ve implemented it correctly.

Then don't apply it until you understand it. Take your time with it.

bilalSerwan commented:
> This [commit](https://github.com/bilalSerwan/ZekirApp/pull/3/commits/66424232b4da2c094803320883840e20ee19ddef) is empty.

Sorry about that; I don't know why the commit is empty.

RoonyShekho09 commented:
> Sorry about that; I don't know why the commit is empty.

It's empty because you didn't add the changed files, make sure you've checked all the files:

<img width="376" alt="Screenshot 2024-08-26 at 11 16 12 AM" src="https://github.com/user-attachments/assets/95c0fa1c-ca1b-487d-a2fc-bf9468194c5c">

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/presentation/ZekirScreen.kt:
```diff
@@ -0,0 +1,121 @@
+fun ZekirScreen(
+    categoryId: Int,
+    navController: NavController,
+    zekirCategoryViewModel: ZekirCategoryViewModel,
```
Each screen should have its own view model, you shouldn't mix different view models with each other.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/appData/SingletonModule.kt:
```diff
@@ -0,0 +1,24 @@
+
+@Module
+@InstallIn(SingletonComponent::class)
+object SingletonModule {
```
SingletonModule doesn't mean anything. Name it AppModule. Also this has nothing to do with data, so don't place it in the data package.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/appData/SingletonModule.kt:
```diff
@@ -0,0 +1,24 @@
+
+@Module
+@InstallIn(SingletonComponent::class)
+object SingletonModule {
```
Actually there is no need for any of this, ZekirCategorySingleton and ZekirSingleton are already singletons, because in Kotlin any object is a singleton. So no need to provide a singleton instance here.

RoonyShekho09 replied:
Please understand the theory behind dependency injection first, after that learn how it's applied in real applications.

bilalSerwan committed:
Refactoring code and modifying the location and structure of files

bilalSerwan committed:
Separating view models for each screen and refactoring the ViewModel code for each screen

bilalSerwan committed:
Removing unnecessary imports and eliminating the unnecessary context parameter in FavoriteScreenViewModel

bilalSerwan committed:
more Refactoring

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/di/AppModule.kt:
```diff
@@ -1,24 +1,27 @@
 
     @Provides
-    @Singleton
     fun provideZekirCategorySingleton(): ZekirCategorySingleton {
```
As I mentioned before, objects are already singletons in Kotlin. No need to provide them. Remove both of the functions.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/di/ZekirSingleton.kt:
```diff
@@ -1,6 +1,7 @@
+import com.fastlink.zekrapp.appData.zekirData
 import javax.inject.Singleton
 
 @Singleton
```
An object is already a singleton, no need for this annotation.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/navigation/Navigation.kt:
```diff
@@ -0,0 +1,32 @@
+    navController: NavController,
+) {
+    NavHost(
+        navController = navController as NavHostController,
```
If you want it as NavHostController then just make the navController type in the parameter as a NavHostController

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/ui/favoriteScreen/FavoriteScreen.kt:
```diff
@@ -1,4 +1,4 @@
-package com.fastlink.zekrapp.presentation
```
UI is part of the presentation layer.

bilalSerwan replied:
We need to create a presentation layer? and move both the navigation and UI to it

RoonyShekho09 replied:
No you don't need it. But I thought you moved the ui out of the presentation layer. Keep it like that if you deleted the presentation layer.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/ui/utils/ZekirCategoryViewModelInterface.kt:
```diff
@@ -0,0 +1,5 @@
+package com.fastlink.zekrapp.ui.utils
+
+interface ZekirCategoryViewModelInterface {
```
Just because you have a function used in two places doesn't mean that you should create an interface for it. Remove this.

bilalSerwan replied:
I didn't create this interface for the updating function. We could implement this function in both ViewModels without an interface. However, I created this interface because, in the `CategoryItemCard` file we have `CategoryItem` composable function, we have a `viewModel` parameter that must accept both `HomeScreenViewModel` and `FavoriteScreenViewModel` since we use this composable in both the `favorite screen` and the `home screen`. There are two ways to handle this: either create a sealed class and extend it with both ViewModels, or create an interface, implement it in both ViewModels, and set the `viewModel` parameter type to this interface to accept both `HomeScreenViewModel` and `FavoriteScreenViewModel`

RoonyShekho09 replied:
That's not how you make your composables reusable in different screens. Rather you should have a callback function in the parameter that handles what will happen when the icon is clicked.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/ui/zekirScreen/ZekirScreen.kt:
```diff
@@ -25,42 +25,39 @@ import androidx.compose.ui.Modifier
     val scaffoldState = rememberScaffoldState()
     val clipboardManager = LocalClipboardManager.current
-    val zekirCategory = zekirCategoryViewModel.getZekirCategoryById(categoryId)
+    val zekirCategory = zekirScreenViewModel.zekirCategory.value!!
```
Don't use the !! if you want to avoid null exceptions.

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/ui/zekirScreen/ZekirScreenViewModel.kt:
```diff
@@ -1,27 +1,45 @@
     val zekirCounter = mutableIntStateOf(0)
     var zekirs: List<ZekirModel> = emptyList()
         private set
+    var zekirCategory: MutableState<ZekirCategoryModel?> = mutableStateOf(null)
```
Why are you making it null, this is unnecessary and makes crashes easier to happen.

bilalSerwan replied:
Initially, I used the lateinit keyword, but it caused the app to crash. So, I changed the type to accept null values to prevent crashes. Now, I have modified the code to this:

` var zekirCategory: MutableState<ZekirCategoryModel> =
        mutableStateOf(ZekirCategoryModel(1, "", ""))`

Is this better, or should I change it again?

RoonyShekho09 replied:
Keep it like this but add default values for the variables in the data class instead of adding values here.

bilalSerwan committed:
add default value to data class

bilalSerwan committed:
deleting  @Singleton notations and provide functions

bilalSerwan committed:
changing type of navController from NavController to NavHostController

bilalSerwan committed:
Delete the interface, replace the viewModel parameter with onClicked, modify the code in HomeScreen and FavoriteScreen, and change the name of the CategoryCard Composable function from CategoryItem to ZekirCategoryCard

bilalSerwan committed:
refactoring code of ZekirScreen and ZekirScreenViewModel to zekirCategory don't accept null value

bilalSerwan committed:
Removing imports that we don't use it

RoonyShekho09 commented on app/src/main/java/com/fastlink/zekrapp/ui/utils/CategoryItemCard.kt:
```diff
@@ -28,11 +28,11 @@ import com.fastlink.zekrapp.appData.model.ZekirCategoryModel
     navController: NavController,
-    viewModel: ZekirCategoryViewModelInterface = hiltViewModel()
-) {
+    onClicked: () -> Unit
```
Specify on clicked on what.

bilalSerwan committed:
Renaming Clicked parameter in ZekirCategoryCard from onCliced to onFavoriteIconClicked

bilalSerwan commented:
thank you so much kak @RoonyShekho09  for Reviewing
